### DISCUSSION TITLE
Client component inside Server Component not rerendering

### DISCUSSION THREAD

**Discussion by F-Moody** [vercel/next.js#62635] [Help]
### Summary

I have a page with a dashboard which is a server component :

```
import PayoutsTableHandler from './payoutsTableHandler'
import { getServerSession } from 'next-auth'
import { authOptions } from '../../../../../pages/api/auth/[...nextauth]'
import { fetchServerSide } from '../../../../../apiUtils/utils'
import {
  GET_HISTORY_FILTERS_PAYOUTS,
  GET_HISTORY_TABLE,
} from '../../../../../apiUtils/costants'

const PayoutTableProvider = async () => {
  const session = await getServerSession(authOptions)

  const historyFilters = await fetchServerSide({
    token: session?.user?.access_token,
    endpoint: GET_HISTORY_FILTERS_PAYOUTS,
  })

  const historyTable = await fetchServerSide({
    token: session?.user?.access_token,
    endpoint: GET_HISTORY_TABLE,
    cache: 'no-cache',
    tags: ['payouts_table'],
  })
  const onFilterChange = async ({ dates: { from, to }, status, page }) => {
    'use server'
    const statusSearchParam = status.slug !== 0 ? status.slug : ''
    const formattedUtcFrom = from ? from.toISOString() : ''
    const formattedUtcTo = to ? to.toISOString() : ''

    return await fetchServerSide({
      token: session?.user?.access_token,
      endpoint: `${GET_HISTORY_TABLE}?status=${statusSearchParam}&page=${page}&date_from=${formattedUtcFrom}&date_to=${formattedUtcTo}`,
    })
  }

  return (
    <PayoutsTableHandler
      filters={historyFilters}
      data={historyTable}
      onFilterChange={onFilterChange}
    />
  )
}

export default PayoutTableProvider
```

Everything seems to be working fine, i can filter with the server action and get the filtered data on the table ( with a useState hook ) and so on.

The problem comes when i open a sideModal ( client component ) that is related with one row (so one element) of the table.

In this sideModal i can upload a document (client component ) and once i've done that i would like to close the modal and revalidate the table fetching the latest data with the status change ( i've uploaded a component ).

This is the client component that uploads a document in the sideModal

```
import { payoutTypes } from '../../constants'
import { useDropzone } from 'react-dropzone'
import ButtonCVA from 'components/ui/buttonCVA'
import { useSidePanelContext } from '../../../../../context/sidePanelContextProvider'
import {
  postAuthenticatedAPI,
  putAuthenticatedApi,
} from '../../../../../apiUtils/utils'
import {
  GET_INVOICE_DATA,
  POST_FILE_UPLOAD_SEEDJ2,
} from '../../../../../apiUtils/costants'
import { useCallback, useState } from 'react'
import { useMessageContext } from '../../../../../context/messageContextProvider'
import { useSession } from 'next-auth/react'
import Image from 'next/image'
import { revalidateTable } from '../../serverActions'

const PayoutFooter = ({ type, data }) => {
  switch (type) {
    case payoutTypes.PAID:
      return <SimpleFooter />

    case payoutTypes.REQUIRED:
      return <UploadFooter data={data} />

    case payoutTypes.REJECTED:
      return <UploadFooter data={data} rejected />

    case payoutTypes.PROCESSING:
      return <SimpleFooter />

    default:
      return null
  }
}

export default PayoutFooter

const UploadFooter = ({ data, rejected }) => {
  const { closeSidePanel } = useSidePanelContext()
  const { setMessageState } = useMessageContext()
  const [loading, setIsLoading] = useState(false)
  const { data: session } = useSession()

  const onUploadInvoice = useCallback(async (acceptedFiles) => {
    let formData = new FormData()
    const file = acceptedFiles[0]
    formData.append('file', file)

    const onSuccess = () => {
      revalidateTable()
      setMessageState({
        type: 'success',
        message: {
          message: 'Invoice uploaded',
        },
      })
      window?.location.reload()
      closeSidePanel()
    }

    const onError = (res) => {
      console.error(res)
      setMessageState({
        type: 'error',
        message: {
          message: 'Something went wrong when uploading the track',
        },
      })
    }

    try {
      setIsLoading(true)
      const id = await postAuthenticatedAPI({
        payload: formData,
        token: session?.user?.access_token,
        endpoint: POST_FILE_UPLOAD_SEEDJ2,
        formData: true,
      })
      if (id) {
        const res = await putAuthenticatedApi({
          payload: { invoice: id },
          token: session?.user?.access_token,
          endpoint: GET_INVOICE_DATA(data.id),
        })
        onSuccess(res)
      }
    } catch (error) {
      onError(error)
    } finally {
      setIsLoading(false)
    }
  }, [])

  const { getRootProps } = useDropzone({
    accept: {
      'application/pdf': ['.pdf'],
      'image/jpg': ['.jpg', '.jpeg', '.png'],
      'application/vnd.ms-excel': ['.xls', '.xlsx'],
    },
    multiple: false,
    noDrag: true,
    onDrop: onUploadInvoice,
  })
  return (
    <div className="flex flex-col items-center gap-4">
      <ButtonCVA
        {...getRootProps({ className: 'dropzone' })}
        variant={'white'}
        size={'medium'}
        className={'w-fit'}
        srcImg="/images/uploadIconBlack.svg"
        imgPosition={'left'}
        imgClassName="w-5 h-5"
        disabled={loading}
      >
        {loading ? (
          <Image
            src="/images/loading.svg"
            alt="loading"
            height="24"
            width="25"
            className="h-[24px] w-[24px] animate-spin "
          />
        ) : (
          `Upload ${rejected ? 'new' : ''} invoice`
        )}
      </ButtonCVA>

      <p className="cursor-pointer text-zinc-50" onClick={closeSidePanel}>
        Go back
      </p>
    </div>
  )
}

const SimpleFooter = () => {
  const { closeSidePanel } = useSidePanelContext()

  return (
    <ButtonCVA
      variant={'white'}
      size={'large'}
      className={'m-auto mt-4'}
      onClick={closeSidePanel}
    >
      Close
    </ButtonCVA>
  )
}
```

To revalidate the tag of the ssr fetch call i use a server action created in a separate file that i call on the onSuccess callback

```
'use server'
import { revalidateTag } from 'next/cache'

export const revalidateTable = () => {
  revalidateTag('payouts_table')
}

```
Now, if i use router.refresh() also in combination with router.push() on the onSuccess callback i cannot refresh the page and i need to manually refresh the page to get the latest data. Now as you can see in the code i need to do a window?.location.reload() which is ugly for me and it's working.

Any idea why here the router.refresh() might not be working?

I also tried putting it inside a startTransition hook but nothing.

I also tried a route handler that revalidates the whole path, waiting for the async call and then using `router.refresh()` but with no luck

**Comment by icyJoseph** [comment]
Hi,

Totally shooting in the dark here, but could you make `revalidateTable` async, and also, await it on your `onSuccess` call? router.refresh should indeed apply to the current route... otherwise, try to reduce this to a minimal reproduction repository, where we could inspect further.

**Reply by F-Moody** [reply]
Hye @icyJoseph , thanks for the answer.
I can't promise to make it into a reproduction repository but this is indeed strange.
How would you make `revalidateTable` async?
I mean, what should i `await` inside ?

And it's strange that as it is it's working with a location reload but not with a `router.refresh()`

Could it be a development environment issue ? Haven't tested on vercel yet

**Reply by icyJoseph** [reply]
Hi,

The thing is that, a server action runs server side, so there'll be a network call from the client to your server, and as you have it, your code just continues through synchronously.

Just label the function async, `await` it on the onSuccess handler, and give it a go, that'd also have the correct semantics, like, `wait for this function to run server side`, even though the work done within it is purely synchronous. I guess there's a race condition where window reload makes it look like it works though. 

Caveat that I haven't had time to read all your code and what's happening. There are many moving parts.

**Reply by F-Moody** [reply]
`      await revalidateTable()
`

and 

```

'use server'
import { revalidateTag } from 'next/cache'

export const revalidateTable = async () => {
  revalidateTag('payouts_table')
}
```


But still nothing, i have to use location.reload. 

It's driving me crazy

**Comment by F-Moody** [comment]
Ok, little update, using some console i see that the ssr components rerenders with the correct data but the children that receives the data as prop don't.

In this case the children is 

```
return (
    <PayoutsTableHandler
      filters={historyFilters}
      data={historyTable}
      onFilterChange={onFilterChange}
    />
  )
```

and the `historyTable` has the updated values but the `PayoutsTableHandler` doesn't rerender and it doesn't display the new data then

**Reply by icyJoseph** [reply]
How have you concluded that it doesn't render? Could it be that the prop is passed to a useState and it never syncs again? Happy to help debugging this

**Reply by F-Moody** [reply]
Because i followed a naive approach.

Set up a console.log on the parent server components and i see that after i revalidate i see the updated data.

Then i set up a console.log on che client component that is the child of it and i see that no console.log is run, so it's like it's not rerendering and so it's not displaying the new data.

Basically the structure is 

Server Component fetches data => pass it down to client component that displays data

After the revalidation

Server components is fetching the new data and it's running again but it's not rerendering his children with the new data


```
'use client'
import PayoutsTableFilters from './payoutsTableFilters'
import PayoutsTable from './payoutsTable'
import { useEffect, useState } from 'react'
import Pagination from 'components/ui/pagination'

const PayoutsTableHandler = ({ onFilterChange, data, filters }) => {
  const [filteredTableData, setFilteredTableData] = useState(data)
  const [filtersState, setFiltersState] = useState({
    status: 0,
    page: 1,
    dates: { from: 0, to: 0 },
  })
  const onStatusFilterChange = async (status) => {
    setFiltersState((filtersState) => ({
      ...filtersState,
      status,
    }))
    setFilteredTableData(await onFilterChange({ ...filtersState, status }))
  }

  const onPageNumFilterChange = async (page) => {
    setFiltersState((filtersState) => ({
      ...filtersState,
      page,
    }))
    setFilteredTableData(await onFilterChange({ filtersState, page }))
  }

  const onDateFilterChange = async ({ from, to }) => {
    setFiltersState((filtersState) => ({
      ...filtersState,
      dates: { from, to },
    }))
    setFilteredTableData(
      await onFilterChange({ ...filtersState, dates: { from, to } })
    )
  }

  useEffect(() => {
    setFilteredTableData(data)
  }, [data])

  return (
    <div className="flex flex-col gap-5">
      <PayoutsTableFilters
        filters={filters}
        onStatusChange={onStatusFilterChange}
        onDatesChange={onDateFilterChange}
      />
      <PayoutsTable data={filteredTableData?.results} />
      <Pagination
        pages={filteredTableData?.page_count}
        currentPage={1}
        maxPageNumberLimit={filteredTableData?.page_count}
        minPageNumberLimit={0}
        onPageChange={onPageNumFilterChange}
      />
    </div>
  )
}

export default PayoutsTableHandler
```
This is the child component, as you can see i also use a `useEffect` to sync up the data but the problem is that the whole component is not running again

**Reply by icyJoseph** [reply]
Do you see a network request, in your browser Dev tools, like ?rsc=, after the revalidation?

**Reply by F-Moody** [reply]
Yes, unfortunatly i cannot create a demo since it's company code but i see the payload requested and i see the console log running on the server 

<img width="601" alt="Screenshot 2024-02-28 alle 20 08 54" src="https://github.com/vercel/next.js/assets/31626129/80d46374-b710-405a-b343-f120a6701f4f">

**Reply by icyJoseph** [reply]
Mmmm 🤔 I don't know then, if the revalidation is sending a new tree state, then it has to be something in the client Tree that's skipping the render, but I can't see what

You could use https://rsc-parser.vercel.app/ to try and see what's being sent by the server. Just paste your rsc payload there.

**Reply by F-Moody** [reply]
Do you know how can i access the rsc payload ? i don't see it in the browser console

**Reply by icyJoseph** [reply]
Right that might be a clue, the browser only shows the payloads if you unwrap them, like via .json() or .text(), otherwise it doesn't show them.

So for some reason in your implementation, which also involves the dependencies of course, the RSC payload is not being consumed. 🤔

**Reply by F-Moody** [reply]
Quick update, after struggling with it, suddently it was working. Please don't ask me why because i cannot understand, the final version is the same as the first one basically.

 ```
const onSuccess = async () => {
      await revalidateTable()
      setMessageState({
        type: 'success',
        message: {
          message: 'Invoice uploaded',
        },
      })
      startTransition(() => {
        router.refresh()
        closeSidePanel()
      })
    }
```

Where `revalidateTable` is a server action.

Thanks for all the support anyway!
